src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java

```diff
@@ -156,56 +211,275 @@ public void leaveToken(DetailAST ast) {
+     * @return number of conditional operators.
+     * @see <a href="http://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.23">
+     * Java Language Specification, &sect;15.23</a>
+     * @see <a href="http://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.24">
+     * Java Language Specification, &sect;15.24</a>
+     * @see <a href="http://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.25">
+     * Java Language Specification, &sect;15.25</a>
+     */
+    private static int countConditionalOperators(DetailAST ast) {
+        int number = 0;
+        for (DetailAST child = ast.getFirstChild(); child != null;
+                child = child.getNextSibling()) {
+            final int type = child.getType();
+            if (type == TokenTypes.LOR || type == TokenTypes.LAND) {
+                number++;
+            }
+            else if (type == TokenTypes.QUESTION) {
+                number += 2;
+            }
+            number += countConditionalOperators(child);
+        }
+        return number;
     }
 
     /**
-     * Pops a value off the stack and makes it the current value.
-     * @return pop a value off the stack and make it the current value
+     * Finds a leaf, which is the most distant from the root.
+     * @param ast the root of tree.
+     * @return the leaf.
      */
-    private BigInteger popValue() {
-        currentValue = valueStack.pop();
-        return currentValue;
+    private static DetailAST getLastToken(DetailAST ast) {
+        final DetailAST lastChild = ast.getLastChild();
+        final DetailAST result;
+        if (lastChild.getFirstChild() == null) {
+            result = lastChild;
+        }
+        else {
+            result = getLastToken(lastChild);
+        }
+        return result;
     }
 
-    /** Process the start of the method definition. */
-    private void visitMethodDef() {
-        pushValue();
+    /**
+     * Counts number of case tokens subject to a case group token.
+     * @param ast case group token.
+     * @return number of case tokens.
+     */
+    private static int countCaseTokens(DetailAST ast) {
+        int counter = 0;
+        for (DetailAST iterator = ast.getFirstChild(); iterator != null;
+                iterator = iterator.getNextSibling()) {
+            if (iterator.getType() == TokenTypes.LITERAL_CASE) {
+                counter++;
+            }
+        }
+        return counter;
     }
 
     /**
-     * Process the end of a method definition.
-     *
-     * @param ast the token representing the method definition
+     * A range between tokens. Used to prevent inline ternary operator from being processed
+     * twice.
      */
-    private void leaveMethodDef(DetailAST ast) {
-        final BigInteger bigIntegerMax = BigInteger.valueOf(max);
-        if (currentValue.compareTo(bigIntegerMax) > 0) {
-            log(ast, MSG_KEY, currentValue, bigIntegerMax);
+    private static class ExpressionRange {
```

Comment:
Technically this isn't a range since it only contains end line number and column number. Start would be needed for it to be a range.